In [99]:
!pip install pyswarm

## Contact angle solver Young-Dubré

### Loading in some libraries

In [2]:
import numpy as np      # mathematics and matrices in Python
from pyswarm import pso # particle swarm optimisation

ModuleNotFoundError: No module named 'numpy'

### Defining classes and functions

In [1]:
class Solvent:    
    def __init__(self, g_LW, g_plus, g_minus, angle):
        self.g_LW = g_LW
        self.g_plus = g_plus
        self.g_minus = g_minus
        self.angle = angle
        
    def surface_tension(self):
        return self.g_LW + 2.*np.sqrt(self.g_plus*self.g_minus)
                

def young_dubre(solvent, gamma_surf):
    gs1, gs2, gs3 = gamma_surf
    return 2.*(np.sqrt(gs1*solvent.g_LW)+np.sqrt(gs2*solvent.g_minus)+np.sqrt(gs3*solvent.g_plus))- \
    (1.+np.cos(np.deg2rad(solvent.angle)))*solvent.surface_tension()

class Three_solvent_system:
    def __init__(self, s1, s2, s3):
        self.s1 = s1
        self.s2 = s2
        self.s3 = s3
    
    def objective(self, gamma_surf):
        SSE = sum([
            young_dubre(self.s1, gamma_surf)**2,
            young_dubre(self.s2, gamma_surf)**2,
            young_dubre(self.s3, gamma_surf)**2
        ])
        return SSE


### Solver settings

In [3]:
lb = [0.0, 0.0, 0.0]           # lower bounds of surface props.
ub = [1000.0, 1000.0, 1000.0]  # upper bounds of surface props.

### Contact angle set 1

In [4]:
water = Solvent(21.8, # gamma_LW
                25.5, # gamma_plus
                25.5, # gamma_minus
                31.7) # contact angle  

glycerol = Solvent(34.0, 3.92, 57.4, 40.0)

diiodomethane = Solvent(50.8, 0.01, 0.0, 48.9)

In [5]:
system = Three_solvent_system(water, glycerol, diiodomethane)

In [6]:
pso(system.objective, lb, ub)

Stopping search: Swarm best objective change less than 1e-08


(array([33.77688979,  1.49852944, 45.47309736]), 4.966515481640414e-09)

### Contact angle set 2

In [7]:
water = Solvent(21.8, 25.5, 25.5, 44.2) # contact angle  

ethyleneGlycol = Solvent(29.0, 3.0, 30.1, 84.7)

diiodomethane = Solvent(50.8, 0.01, 0.0, 65.0)

In [8]:
system = Three_solvent_system(water, ethyleneGlycol, diiodomethane)

In [9]:
pso(system.objective, lb, ub)

Stopping search: Swarm best objective change less than 1e-08


(array([17.69039602,  0.        , 61.30077823]), 558.914109124122)

Hierboven kan je zien dat hij convergeert naar je MATLAB systeem maar aan de waarde van je doelfunctie `558.91` zie je dat de Young Dubré vergelijking niet echt goed voldaan is